packages/terra-dropdown-button/src/DropdownButton.jsx

```diff
@@ -57,29 +57,56 @@ class DropdownButton extends React.Component {
     this.handleDropdownRequestClose = this.handleDropdownRequestClose.bind(this);
     this.handleKeyDown = this.handleKeyDown.bind(this);
     this.handleKeyUp = this.handleKeyUp.bind(this);
+    this.setButtonNode = this.setButtonNode.bind(this);
+    this.getButtonNode = this.getButtonNode.bind(this);
+    this.setListNode = this.setListNode.bind(this);
+    this.state = { isOpen: false, isActive: false, openedViaKeyboard: false };
+  }
+
+  setListNode(element) {
+    this.dropdownList = element;
+  }
 
-    this.state = { isOpen: false, isActive: false };
+  setButtonNode(node) {
+    this.buttonNode = node;
+  }
+
+  getButtonNode() {
+    return this.buttonNode;
   }
 
   handleDropdownButtonClick() {
+    if (this.state.isOpen) {
+      this.setState({ openedViaKeyboard: false });
+    }
     this.setState(prevState => ({ isOpen: !prevState.isOpen }));
   }
 
   handleDropdownRequestClose(callback) {
-    this.setState({ isOpen: false }, typeof callback === 'function' ? callback : undefined);
+    const callbackOnRequestClose = typeof callback === 'function' ? callback : undefined;
```

Comment:
Could this be simplified to - 

`this.setState({ isOpen: false, openedViaKeyboard: false, isActive: false }, callback);`

Is there a scenario wherein the argument that is passed in is not a function?

Reply:
I did not find any existing scenario where we are passing callback function to handleDropdownRequestClose() method and there is no callback for onClick is provided in DropdownButton for consumers to pass in callback. so am not sure when it is passed in as non-function.
 I feel this might be added has safety check to ensure callback that will triggered in handleDropdownRequestClose() is a function.

@StephenEsser would like to know your thoughts on callback usage in handleDropdownRequestClose() method

Reply:
Maybe, that could be removed and introduced whenever needed. I don't see its usage in the [original PR](https://github.com/cerner/terra-core/pull/2487) either.

Reply:
Edit : I found that callback which we will be receiving in handleDropdownRequestClose() function is onSelect() function of selected item. So on item selection `onSelect(event, metaData)` will be passed callback function and on outside click `mouse-event` will be passed as callback which is not of a type `function` hence we need this check here before invoking callback.